torchgeo/datasets/utils.py

```diff
@@ -631,6 +635,74 @@ def path_is_vsi(path: Path) -> bool:
     return '://' in str(path) or str(path).startswith('/vsi')
 
 
+def listdir_vsi_recursive(root: Path) -> list[str]:
+    """Lists all files in Virtual File Systems (VSI) recursively.
+
+    Args:
+        root: directory to list. These must contain the prefix for the VSI
+            e.g. '/vsiaz/' or 'az://' for azure blob storage
+                 '/vsizip/' or 'zip://' for zipped archives
+
+    Returns:
+        A list of all file paths matching filename_glob in the root VSI directory or its
+        subdirectories.
+
+    .. versionadded:: #0.6
+    """
+    dirs = [str(root)]
+    files = []
+    while dirs:
+        dir = dirs.pop()
+        try:
+            subdirs = fiona.listdir(dir)
+            dirs.extend([os.path.join(dir, subdir) for subdir in subdirs])
+        except FionaValueError as e:
+            if 'is not a directory' in str(e):
+                # Assuming dir is a file as it is not a directory
+                # fiona.listdir can throw FionaValueError for only two reasons
+                files.append(dir)
+            else:
+                # fiona.listdir can throw FionaValueError for only two reasons
+                # 1. 'is not a directory'
+                # 2. 'does not exist'
+                raise
+    return files
+
+
+def list_directory_recursive(root: Path, filename_glob: str) -> list[str]:
+    """Lists files in directory recursively matching the given glob expression.
+
+    Also supports GDAL Virtual File Systems (VSI).
+
+    Args:
+        root: directory to list. For VSI these will have prefix
+            e.g. /vsiaz/ or az:// for azure blob storage
+        filename_glob: filename pattern to filter filenames
+
+    Returns:
+        A list of all file paths matching filename_glob in the root directory or its
+        subdirectories.
+
+    .. versionadded:: #0.6
+    """
+    files: list[str]
+    if path_is_vsi(root):
+        # Change type to match expected input to filter
+        all_files: list[str] = []
+        try:
+            all_files = listdir_vsi_recursive(root)
+        except FionaValueError:
+            # To match the behaviour of glob.iglob we silently return empty list
+            # for non-existing root.
+            pass
+        # Prefix glob with wildcard to ignore directories
+        files = fnmatch.filter(all_files, os.path.join('*', filename_glob))
```

Comment:
I chose not to use `os.path.join` here because the VSI may be a URL and use / even on Windows where `os.path.join` uses \

Reply:
Ah, yes. I [see it now](https://github.com/microsoft/torchgeo/blob/042b75ea9c884d2743bb46687ee0bd069dfc2a79/torchgeo/datasets/geo.py#L315).

This should be the reason the reason for the failing test on windows. I need to do the same in [listdir_vfs_recursive](https://github.com/microsoft/torchgeo/blob/51a3d64f6f0339cb26707cea2579b952fd725605/torchgeo/datasets/utils.py#L662)

Reply:
Updated it and removed the skipif windows